Add policy-related features tracking in Cilium agent as prometheus metrics

aanm opened the pull request:
Follow up of https://github.com/cilium/cilium/pull/35852 for more features. I've split in multiple commits so it's easier for each sig to review their section.

**Note for reviewers**: Review on a per-commit basis to make sure the feature of your sig is being detected correctly.

### Features Checklist
- [ ] Enable L3 policy (@cilium/sig-policy)
- [ ] Enable Host Network Policies (@cilium/sig-policy)
- [ ] Enable DNS rules (@cilium/sig-policy)
- [ ] Enable HTTP and HTTPHeaderMatches Network Policies (@cilium/sig-policy)
- [ ] Enable other L7 Network Policies (@cilium/sig-policy)
- [ ] Enable Deny Network Policies (@cilium/sig-policy)
- [ ] Enable IngressCIDR Network Policies (@cilium/sig-policy)
- [x] Enable Mutual Auth Network Policies (@cilium/sig-servicemesh)
- [x] Enable TLS and SNI Network Policies (@cilium/sig-servicemesh)
- [ ] Enable Default Deny Network Policies (@cilium/sig-policy)
- [ ] Enable Local Redirect Policies (@cilium/sig-datapath)
- [ ] Enable Service Internal Traffic Policy (@cilium/sig-datapath)
- [x] Enable CiliumEnvoyConfig (@cilium/sig-servicemesh)
- [ ] Enable CiliumNetworkPolicy (@cilium/sig-policy)
- [ ] Enable Host Network Policies (@cilium/sig-policy)
- [ ] Enable ToFQDNs (@cilium/sig-policy)

aanm committed:
metrics/features: enable L3 policy

Track if L3 policies are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable Host Network Policies

Track if Host Network Policies are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable DNS rules

Track if DNS rules are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable HTTP and HTTPHeaderMatches Network Policies

Track if HTTP and HTTPHeaderMatches Network Policies are deployed
in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable other L7 Network Policies

Track if other L7 network policies are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable Deny Network Policies

Track if Deny Network Policies are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable IngressCIDR Network Policies

Track if IngressCIDR Network Policies are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable Mutual Auth Network Policies

Track if Mutual Auth Network Policies are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable TLS and SNI Network Policies

Track if TLS and SNI Network Policies are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable Default Deny Network Policies

Track if Default Deny Network Policies is deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable Local Redirect Policies

Track if LRP is deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable Service Internal Traffic Policy

Track if Services with Traffic Policy="Internal" are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable CiliumEnvoyConfig

Track if CiliumEnvoyConfig and CiliumClusterwideEnvoyConfig are deployed
in Cilium

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
metrics/features: enable CiliumNetworkPolicy

Track if CiliumNetworkPolicy and CiliumClusterwideNetworkPolicy are
deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

giorio94 reviewed (approved):
/lgtm for the trivial clustermesh-related change

tklauser commented on pkg/policy/api/egress.go:
```diff
@@ -363,6 +363,19 @@ func (e *EgressCommonRule) RequiresDerivative() bool {
 	return len(e.ToGroups) > 0
 }
 
+func (in *EgressCommonRule) IsL3() bool {
```
nit: receiver name should be `e` instead of `in` in correspondence with the other `*EgressCommonRule` methods.

tklauser commented on pkg/metrics/features/metrics.go:
```diff
@@ -512,6 +515,20 @@ func NewMetrics(withDefaults bool) Metrics {
 			Name:      "cilium_node_config_enabled",
 		}),
+
+		NPL3Ingested: metric.NewGauge(metric.GaugeOpts{
```
`NPL3Ingested` is defined as a `metric.Counter` but this is instantiated using `metric.NewGauge`. Not sure whether it matters.

aanm replied:
It's odd that it didn't complain, I'll change it to a counter.

aanm replied:
@tklauser it turns out that this should be defined as a metric.Gauge :sweat_smile: I forgot about this.

tklauser replied:
> It's odd that it didn't complain, I'll change it to a counter.

I guess that's because the `metric.Counter` interfaceis a subset of the `metric.Gauge` interface.

> @tklauser it turns out that this should be defined as a metric.Gauge 😅 I forgot about this.

That's fine too. As long as they are consistent :sweat_smile:

tklauser commented on pkg/metrics/features/metrics.go:
```diff
@@ -529,6 +531,20 @@ func NewMetrics(withDefaults bool) Metrics {
 			Name:      "l3_l4_policies_present",
 		}),
+
+		NPHostNPIngested: metric.NewGauge(metric.GaugeOpts{
```
Same here, `NPHostNPIngested` defined as `metric.Counter` but initialized using `metric.NewGauge`.

tklauser commented on pkg/metrics/features/metrics.go:
```diff
@@ -545,6 +547,20 @@ func NewMetrics(withDefaults bool) Metrics {
 			Name:      "host_network_policies_present",
 		}),
+
+		NPDNSIngested: metric.NewGauge(metric.GaugeOpts{
```
And same here and with all other `*Ingested` metrics added in successive commits.

christarazi commented on pkg/metrics/features/policy.go:
```diff
@@ -94,10 +122,16 @@ func ruleType(r api.Rule) RuleFeatures {
+				if len(e.ToFQDNs) > 0 {
+					rf.DNS = true
+				}
+				ruleTypePortRules(&rf, e.ToPorts)
```
So this metric will also count `ToFQDN` rules as well as `DNS` rules? What if the user has only `ToFQDN` rules and not `DNS` or vice versa? A common mistake is to define a `ToFQDN` without a corresponding DNS rule. I'm not sure if we want to expose that at this level though. :thinking:

aanm replied:
@christarazi That's why I wanted someone from the policy sig to review this PR as I was not so sure about this as well. Should we then distinguish between DNS and ToFQDN policies?

christarazi replied:
My feeling is yes we should distinguish between them. Otherwise, it can be misleading. How do others in @cilium/sig-policy feel?

aanm replied:
Done

christarazi replied:
I see that the changes were made in the last commit to add separate counting for FQDN, but don't we need to remove the 

```
if len(e.ToFQDNs) > 0 {
	rf.DNS = true
}			
```

block too?

christarazi replied:
Addressed in https://github.com/cilium/cilium/pull/36203#discussion_r1868327428

mhofstetter commented on pkg/ciliumenvoyconfig/cec_manager.go:
```diff
@@ -325,6 +334,16 @@ func (r *cecManager) updateCiliumEnvoyConfig(
 	oldCECObjectMeta metav1.ObjectMeta, oldCECSpec *ciliumv2.CiliumEnvoyConfigSpec,
 	newCECObjectMeta metav1.ObjectMeta, newCECSpec *ciliumv2.CiliumEnvoyConfigSpec,
 ) error {
+	if oldCECObjectMeta.GetNamespace() == "" {
```
couldn't we get rid of this (edit: `this`= updating any metrics in `updateCiliumEnvoyConfig`) completely? as decreasing and immediately increasing the metric is a "no-op" (as the underlying metric isn't interested in the spec)

aanm committed:
metrics/features: fix NodePortConfig test case

Even if NodePortConfig acceleration is disabled it doesn't mean that all
other configurations are not set. Thus, we can remove this test case.

Signed-off-by: André Martins <andre@cilium.io>

giorio94 commented on pkg/metrics/features/metrics.go:
```diff
@@ -804,6 +830,34 @@ func NewMetrics(withDefaults bool) Metrics {
 		}),
+
+		ACLBClusterMeshEnabled: metric.NewGaugeVecWithLabels(metric.GaugeOpts{
+			Help:      "Clustermesh mode enabled on the agent",
```
I'd suggest slightly changing the description here (and potentially the metric name) to reflect the fact that this is not an agent setting, but rather reflects the characteristics of the clusters this component is connected to. Mentioning the agent is also a bit incorrect, because this metric is also potentially exposed by the Cilium operator and kvstoremesh.

aanm replied:
Solve, I've added a NoopImplementation to avoid blocking ClusterMesh from initializing.

giorio94 replied:
I'd still be in favor of adapting the name/description, as the current version is potentially misleading given that it is not tracking the characteristics of the current cluster, but the connections to remote ones. Maybe just `clustermesh` (or `clustermesh_peers`, `clustermesh_connections`) as name, and `Mode of the active Cluster Mesh connections/peers`, or something along these lines. WDYT?

giorio94 commented on pkg/metrics/features/misc_test.go:
```diff
@@ -297,3 +298,58 @@ func TestCCNP(t *testing.T) {
+				maxClusters: maxClusters,
+
+				expectedMode:        mode,
+				expectedMaxClusters: maxClusters,
```
Nit. Looks redundant, as always configured to the same values as `mode`, `maxCluster`.

giorio94 commented on pkg/metrics/features/misc_test.go:
```diff
@@ -297,3 +298,58 @@ func TestCCNP(t *testing.T) {
+					}
+					metrics.DelClusterMeshConfig(tt.mode, tt.maxClusters)
+					counter, err = metrics.ACLBClusterMeshEnabled.GetMetricWithLabelValues(mode, maxClusters)
+					assert.NoError(t, err)
```
Nit. looks redundant, as it is equivalent to the counter that has already been retrieved above.

aanm replied:
Not really, the value will be different.

giorio94 replied:
Sure, but the `counter` object is always the same one, it is not bound to a specific value.

giorio94 commented on pkg/metrics/features/misc_test.go:
```diff
@@ -297,3 +298,58 @@ func TestCCNP(t *testing.T) {
+			for _, mode := range defaultClusterMeshMode {
+				for _, maxClusters := range defaultClusterMeshMaxConnectedClusters {
+
+					metrics := NewMetrics(true)
```
Is it intentional that the metrics object is initialized at every iteration?

aanm replied:
Yes, it's just to make sure the metrics object is new for each test case.

giorio94 commented on pkg/metrics/features/misc_test.go:
```diff
@@ -297,3 +298,58 @@ func TestCCNP(t *testing.T) {
+					metrics.AddClusterMeshConfig(tt.mode, tt.maxClusters)
+
+					counter, err := metrics.ACLBClusterMeshEnabled.GetMetricWithLabelValues(mode, maxClusters)
+					assert.NoError(t, err)
```
Nit. `require.NoError` is typically more appropriate in case of errors because, if that were to happen, `counter` would be likely nil, and trigger a panic below.

giorio94 commented on pkg/clustermesh/common/remote_cluster.go:
```diff
@@ -122,6 +129,11 @@ func (rc *remoteCluster) releaseOldConnection() {
+	// if rc.config is nil then this remote cluster was already deleted and
+	// it wasn't re-added yet.
+	if rc.config != nil {
+		rc.featureMetrics.DelClusterMeshConfig(ClusterMeshMode(rc.config), rc.featureMetricMaxClusters)
```
I'd move this to line 234, for consistency with the corresponding increasing step. That also prevents the possibility that a decrease is triggered for a configuration for which the corresponding increase operation has not happened (i.e., if the configuration retrieval has failed).

aanm replied:
Ah, it makes more sense, yes.

giorio94 commented on pkg/clustermesh/common/clustermesh.go:
```diff
@@ -270,3 +282,25 @@ func (cm *clusterMesh) ForEachRemoteCluster(fn func(RemoteCluster) error) error
+)
+
+// ClusterMeshMode returns the connection type made to the remote cluster.
+func ClusterMeshMode(rcc *models.RemoteClusterConfig) string {
```
Nit. I'd change the parameter to the actual `CiliumClusterConfig` struct (defined in `pkg/clustermesh/types`), rather than the derived representation used for the Cilium API, for consistency with the other usages.

giorio94 commented on pkg/clustermesh/common/clustermesh.go:
```diff
@@ -270,3 +282,25 @@ func (cm *clusterMesh) ForEachRemoteCluster(fn func(RemoteCluster) error) error
+func ClusterMeshMode(rcc *models.RemoteClusterConfig) string {
+	if rcc == nil || !rcc.Retrieved {
+		return ClusterMeshModeUndetectable
+	}
```
Nit. I'd move as a case of the switch, for consistency and ease of readability.

giorio94 commented on pkg/metrics/features/metrics.go:
```diff
@@ -804,6 +830,34 @@ func NewMetrics(withDefaults bool) Metrics {
+						defaultClusterMeshMaxConnectedClusters...,
+					)
+				}(),
+			},
```
I wonder if it could make sense to break this into a separate metric. The reason being that the mode is specific to the clusters we are connected to, while the max_connected_clusters number of an agent specific setting (all clusters in a clustermesh must use the same value). WDYT?

giorio94 reviewed (changes requested):
I've reviewed the clustermesh-specific commit, and left a few comments inline. Mostly nits, the only one which is more relevant concerns ensuring that metric increase and decrease is always symmetric.

aanm committed:
metrics/features: enable ToFQDN rules

Track if ToFQDN rules are deployed in Cilium

Signed-off-by: André Martins <andre@cilium.io>

giorio94 commented on pkg/clustermesh/common/clustermesh.go:
```diff
@@ -270,3 +282,21 @@ func (cm *clusterMesh) ForEachRemoteCluster(fn func(RemoteCluster) error) error
+func ClusterMeshMode(rcc types.CiliumClusterConfig) string {
+	switch {
+	case rcc.Capabilities.Cached:
+		return ClusterMeshModeClusterMeshAPIServer
```
`ClusterMeshModeClusterMeshAPIServer` and `ClusterMeshModeKVStoreMesh` are inverted.

giorio94 commented on operator/cmd/root.go:
```diff
@@ -276,6 +277,8 @@ var (
 			// features are enabled on the operator.
 			features.Cell,
+
+			featuresNoop.Cell,
```
One of the two variants of the metrics should be provided in kvstoremesh as well: https://github.com/cilium/cilium/blob/pr%2Faanm%2Ffeatures-3/clustermesh-apiserver/kvstoremesh/cells.go#L22-L55

Just a note that providing the noop implementation in kvstoremesh would potentially mean having a partial view of the connections from the local cluster when kvstoremesh is enabled, as all the agents would report kvstoremesh, and it would not be possible to distinguish whether the remote clusters actually use the clustermesh-apiserver or an etcd cluster (assuming that's of interest here).

giorio94 commented on pkg/clustermesh/common/clustermesh.go:
```diff
@@ -44,6 +44,14 @@ type Configuration struct {
 	Metrics Metrics
+
+	// FeatureMetrics will track which features are enabled with in clustermesh.
+	FeatureMetrics ClusterMeshMetrics
```
This variable is currently never initialized, neither in production code nor in tests.

aanm commented on pkg/metrics/features/policy.go:
```diff
@@ -207,16 +214,16 @@ func ruleType(r api.Rule) RuleFeatures {
+			if !(rf.allFeaturesPortRules() && rf.ToFQDNs) {
 				if len(e.ToFQDNs) > 0 {
-					rf.DNS = true
+					rf.ToFQDNs = true
```
@christarazi you mean this one?

christarazi replied:
Yep!

learnitall reviewed (approved):
API changes LGTM

giorio94 reviewed (approved):
Approving again the only remaining clustermesh-related change.
